<a href="https://colab.research.google.com/github/yashfirkedata/DL-Patient-Survival-Prediction/blob/main/patient_survival_prediction_modelling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Modelling on Patient Survival Prediction**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Activation

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
data = pd.read_csv("/content/drive/MyDrive/Datasets/psp_modelling_data.csv")

In [4]:
data.head()

hospital_death   age        bmi  elective_surgery  height  \
0               0  68.0  22.730000                 0   180.3   
1               0  77.0  27.420000                 0   160.0   
2               0  67.0  27.560000                 0   190.5   
3               0  46.0  25.845717                 0   167.6   
4               0  87.0  21.963763                 0   180.3   

   pre_icu_los_days  weight  apache_2_diagnosis  apache_3j_diagnosis  \
0          0.541667    73.9               113.0               502.01   
1          0.927778    70.2               108.0               203.01   
2          0.000694   100.0               301.0               403.01   
3          0.000000    72.6               108.0               203.01   
4          5.046528    71.4               113.0               501.05   

   apache_post_operative  ...  isin_Cardiovascular.1  isin_Gastrointestinal.1  \
0                      0  ...                   True                    False   
1                      0  ...                  False                    False   
2                      0  ...                  False                    False   
3                      0  ...                  False                    False   
4                      0  ...                   True                    False   

   isin_Haematologic  isin_Metabolic.1  isin_Neurologic  \
0              False             False            False   
1              False             False            False   
2              False             False             True   
3              False             False            False   
4              False             False            False   

   isin_Renal/Genitourinary  isin_Respiratory.1  isin_Trauma.1  \
0                     False               False          False   
1                     False                True          False   
2                     False               False          False   
3                     False                True          False   
4                     False               False          False   

   isin_Undefined Diagnoses  isin_Undefined diagnoses  
0                     False                     False  
1                     False                     False  
2                     False                     False  
3                     False                     False  
4                     False                     False  

[5 rows x 136 columns]

In [5]:
# separating target and predictor variables
data_y = data['hospital_death']
data_x = data.drop('hospital_death',axis=1)

In [6]:
data_y.value_counts()

hospital_death
0    48461
1     4652
Name: count, dtype: int64

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data_x, data_y,  test_size=0.2, shuffle = True, random_state = 8)

In [8]:
X_train.shape, y_train.shape

((42490, 135), (42490,))

In [9]:
pd.set_option('display.max_columns', None)
X_train.head()

age        bmi  elective_surgery  height  pre_icu_los_days  weight  \
28981  69.0  31.972318                 1   170.0          2.662500    92.4   
34168  43.0  42.292992                 0   167.6          0.169444   118.8   
17273  30.0  21.527821                 0   152.4         -0.004167    50.0   
19501  70.0  30.782267                 0   170.8          0.227083    89.8   
45315  65.0  41.602450                 1   165.1          0.946528   113.4   

       apache_2_diagnosis  apache_3j_diagnosis  apache_post_operative  \
28981               308.0              1207.01                      1   
34168               113.0               501.05                      0   
17273               123.0               702.01                      0   
19501               110.0               110.01                      0   
45315               308.0              1207.01                      1   

       arf_apache  bun_apache  creatinine_apache  gcs_eyes_apache  \
28981           0         7.0               0.60              3.0   
34168           0         9.0               0.51              4.0   
17273           0        18.0               0.76              4.0   
19501           0        14.0               0.93              4.0   
45315           0        11.0               0.87              3.0   

       gcs_motor_apache  gcs_unable_apache  gcs_verbal_apache  glucose_apache  \
28981               6.0                  0                3.0           127.0   
34168               6.0                  0                5.0           184.0   
17273               6.0                  0                5.0           298.0   
19501               6.0                  0                5.0            79.0   
45315               6.0                  0                1.0           167.0   

       heart_rate_apache  hematocrit_apache  intubated_apache  map_apache  \
28981               91.0               28.2                 1        57.0   
34168              101.0               42.2                 0       134.0   
17273              139.0               40.1                 0        71.0   
19501               92.0               39.1                 0       117.0   
45315               94.0               25.0                 1        62.0   

       resprate_apache  sodium_apache  temp_apache  ventilated_apache  \
28981             29.0          138.0         36.8                  1   
34168             39.0          132.0         35.8                  1   
17273             29.0          140.0         36.7                  0   
19501             26.0          134.0         36.0                  0   
45315             26.0          137.0         35.8                  1   

       wbc_apache  d1_diasbp_max  d1_diasbp_min  d1_diasbp_noninvasive_max  \
28981       14.90           67.0           41.0                       67.0   
34168       13.75          123.0           78.0                      123.0   
17273       12.70           81.0           57.0                       81.0   
19501        8.00          114.0           63.0                      114.0   
45315       14.90          101.0           52.0                      101.0   

       d1_diasbp_noninvasive_min  d1_heartrate_max  d1_heartrate_min  \
28981                       41.0              91.0              80.0   
34168                       78.0             100.0              41.0   
17273                       57.0             134.0              93.0   
19501                       63.0              90.0              72.0   
45315                       52.0              94.0              60.0   

       d1_mbp_max  d1_mbp_min  d1_mbp_noninvasive_max  d1_mbp_noninvasive_min  \
28981        73.0        57.0                    73.0                    57.0   
34168       134.0        89.0                   134.0                    89.0   
17273       101.0        71.0                   101.0                    71.0   
19501       117.0        72.0                   117.0                 

###**Standardization**

In [10]:
from sklearn.preprocessing import MinMaxScaler
# creating an instance for MinMaxScaler
scaler = MinMaxScaler()
X_train_std = scaler.fit_transform(X_train)
X_test_std = scaler.transform(X_test)

In [11]:
# storing the normalized data into a new dataframe
X_train_new = pd.DataFrame(X_train_std, columns= X_train.columns)
X_test_new = pd.DataFrame(X_test_std, columns= X_test.columns)

In [12]:
X_test_new.head()

age       bmi  elective_surgery    height  pre_icu_los_days    weight  \
0  0.794521  0.337382               0.0  0.733002          0.004253  0.457259   
1  0.452055  0.383497               0.0  0.303134          0.004367  0.310448   
2  0.136986  0.235238               0.0  0.652509          0.005812  0.307395   
3  0.780822  0.639042               0.0  0.565165          0.064935  0.695115   
4  0.520548  0.243601               0.0  0.912828          0.058276  0.421303   

   apache_2_diagnosis  apache_3j_diagnosis  apache_post_operative  arf_apache  \
0            0.077295             0.002381                    0.0         0.0   
1            0.053140             0.002857                    0.0         0.0   
2            0.082126             0.239995                    0.0         0.0   
3            0.057971             0.190473                    0.0         0.0   
4            0.057971             0.190492                    0.0         0.0   

   bun_apache  creatinine_apache  gcs_eyes_apache  gcs_motor_apache  \
0    0.089431           0.048713         1.000000               1.0   
1    0.008130           0.035846         1.000000               1.0   
2    0.056911           0.050551         1.000000               1.0   
3    0.447154           0.168199         1.000000               1.0   
4    0.048780           0.036765         0.333333               0.8   

   gcs_unable_apache  gcs_verbal_apache  glucose_apache  heart_rate_apache  \
0                0.0                1.0        0.176880           0.209459   
1                0.0                1.0        0.175094           0.445946   
2                0.0                1.0        0.180454           0.540541   
3                0.0                1.0        0.214401           0.175676   
4                0.0                0.0        0.307307           0.459459   

   hematocrit_apache  intubated_apache  map_apache  resprate_apache  \
0           0.846591               0.0     0.17500         0.142857   
1           0.696023               0.0     0.47500         0.464286   
2           0.775568               0.0     0.54375         0.517857   
3           0.420455               1.0     0.06250         0.107143   
4           0.826705               1.0     0.51250         0.571429   

   sodium_apache  temp_apache  ventilated_apache  wbc_apache  d1_diasbp_max  \
0       0.560976     0.592105                0.0    0.109131       0.403361   
1       0.414634     0.486842                0.0    0.316927       0.537815   
2       0.414634     0.526316                0.0    0.238976       0.344538   
3       0.414634     0.539474                1.0    0.165033       0.327731   
4       0.463415     0.607895                1.0    0.142539       0.226891   

   d1_diasbp_min  d1_diasbp_noninvasive_max  d1_diasbp_noninvasive_min  \
0       0.519481                   0.403361                   0.519481   
1       0.740260                   0.537815                   0.740260   
2       0.402597                   0.344538                   0.402597   
3       0.428571                   0.327731                   0.428571   
4       0.779221                   0.226891                   0.779221   

   d1_heartrate_max  d1_heartrate_min  d1_mbp_max  d1_mbp_min  \
0          0.210084          0.429530    0.314516    0.511111   
1          0.277311          0.483221    0.451613    0.711111   
2          0.420168          0.382550    0.314516    0.677778   
3          0.151261          0.382550    0.379032    0.311111   
4          0.327731          0.489933    0.209677    0.711111   

   d1_mbp_noninvasive_max  d1_mbp_noninvasive_min  d1_resprate_max  \
0                0.322314                0.511111         0.141026   
1                0.462810                0.711111         0.205128   
2                0.322314                0.677778         0.243590   
3                0.388430                0.311111         0.128205   
4                0.214876                0.711111 

#**ML Model 1: XGBoostClassifier**

In [13]:
from xgboost import XGBClassifier

model = XGBClassifier(use_label_encoder=False, eval_metric='logloss')
model.fit(X_train_new, y_train)

y_pred = model.predict(X_test_new)
y_pred_proba = model.predict_proba(X_test_new)[:, 1]

In [14]:
# Calculate metrics
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, confusion_matrix

accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
roc_auc = roc_auc_score(y_test, y_pred_proba)
conf_matrix = confusion_matrix(y_test, y_pred)

# Print metrics
print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1 Score: {f1}")
print(f"ROC AUC Score: {roc_auc}")
print(f"Confusion Matrix:\n{conf_matrix}")


Accuracy: 0.9229031347077097
Precision: 0.592814371257485
Recall: 0.3256578947368421
F1 Score: 0.4203821656050955
ROC AUC Score: 0.8802435901952389
Confusion Matrix:
[[9507  204]
 [ 615  297]]


#**ML Model 2: Logistic Regression**

In [15]:
from sklearn.linear_model import LogisticRegression

# Initialize the Logistic Regression model
log_reg = LogisticRegression()

# Train the model
log_reg.fit(X_train_new, y_train)

# Make predictions
y_pred = log_reg.predict(X_test_new)
y_pred_proba = log_reg.predict_proba(X_test_new)[:, 1]

# Calculate metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
roc_auc = roc_auc_score(y_test, y_pred_proba)
conf_matrix = confusion_matrix(y_test, y_pred)

# Print metrics
print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1 Score: {f1}")
print(f"ROC AUC Score: {roc_auc}")
print(f"Confusion Matrix:\n{conf_matrix}")

Accuracy: 0.9240327591075967
Precision: 0.6271186440677966
Recall: 0.28399122807017546
F1 Score: 0.3909433962264151
ROC AUC Score: 0.8720621351803978
Confusion Matrix:
[[9557  154]
 [ 653  259]]


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


#**ML Model 3: DecisionTreeClassifier**

In [16]:
from sklearn.tree import DecisionTreeClassifier

tree_model = DecisionTreeClassifier()

# Train the model
tree_model.fit(X_train_new, y_train)

# Make predictions
y_pred = tree_model.predict(X_test_new)
y_pred_proba = tree_model.predict_proba(X_test_new)[:, 1]

# Calculate metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
roc_auc = roc_auc_score(y_test, y_pred_proba)
conf_matrix = confusion_matrix(y_test, y_pred)

# Print metrics
print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1 Score: {f1}")
print(f"ROC AUC Score: {roc_auc}")
print(f"Confusion Matrix:\n{conf_matrix}")

Accuracy: 0.8743292855125671
Precision: 0.3025210084033613
Recall: 0.35526315789473684
F1 Score: 0.3267776096822996
ROC AUC Score: 0.6391700404858299
Confusion Matrix:
[[8964  747]
 [ 588  324]]


**It seems like XGBoost has highest ROC-AUC score so we will try hyperparameter tuning**

## **Hyperparameter Tuning - XGBoost**

In [17]:
from sklearn.model_selection import KFold, GridSearchCV

# K-Fold cross-validation
gkf = KFold(n_splits=3, shuffle=True, random_state=42).split(X=X_train_new, y=y_train)

# Parameter grid for XGBoost
params = {
    'booster': ["gbtree"],
    'learning_rate': [0.1],
    'n_estimators': range(100, 500, 100)
}

xgb_estimator = XGBClassifier()

# Grid search for best parameters
gsearch = GridSearchCV(
    estimator=xgb_estimator,
    param_grid=params,
    scoring='roc_auc',
    n_jobs=-1,
    cv=gkf
)

xgb_model = gsearch.fit(X=X_train, y=y_train)
best_params, best_score = gsearch.best_params_, gsearch.best_score_

(best_params, best_score)

/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/fork_exec.py:38: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid = os.fork()


({'booster': 'gbtree', 'learning_rate': 0.1, 'n_estimators': 100},
 0.886000685077824)

## **Final ML model tuned**

In [20]:
main_ml_model = XGBClassifier(n_estimators=100,
    booster="gbtree",
    learning_rate=0.1,
    n_jobs=-1)

main_ml_model.fit(X_train_new, y_train)

# Make predictions
y_pred = main_ml_model.predict(X_test_new)
y_pred_proba = main_ml_model.predict_proba(X_test_new)[:, 1]

# Calculate metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
roc_auc = roc_auc_score(y_test, y_pred_proba)
conf_matrix = confusion_matrix(y_test, y_pred)

# Print metrics
print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1 Score: {f1}")
print(f"ROC AUC Score: {roc_auc}")
print(f"Confusion Matrix:\n{conf_matrix}")

Accuracy: 0.9287395274404594
Precision: 0.679814385150812
Recall: 0.32127192982456143
F1 Score: 0.4363365599404319
ROC AUC Score: 0.8902733064511759
Confusion Matrix:
[[9573  138]
 [ 619  293]]


**Okay so best Machine Learning Model we have settled for is XGBoost Classifer with tuned parameters**

*   ROC AUC Score = 0.8902733064511759



Now lets try Deep Learning modelling